### DISCUSSION TITLE
3d probability distribution

### DISCUSSION THREAD

**Discussion by Negar-Erfanian** [pyvista/pyvista#1904] [Q&A] [answered]
Hi,
I'm very new to pyvista (actually just started working with it today)

I have a 3d probability density which is (N,N,N).

As an example having 

```py
x = np.arange(0, 10, N)
y = np.arange(0, 10, N)
z = np.arange(0, 10, N)
x, y , z = np.meshgrid(x, y, z)

p = density_func(x,y,z)
```

and p is finally (N,N,N).

Now I would like to have a 3d visualization of p but I don't know how to have a clean visualization.
I was wondering if this is the only way to do this?

data = pv.wrap(p)
pv.plot(values, background = 'w',hidden_line_removal = True)

Thanks for your help in advance!

**Comment by banesullivan** [comment, marked as answer]
To make a mesh is simply:

```py
mesh = pv.StructuredGrid(x, y, z)
mesh['p'] = p.ravel(order='f')
```

Here is an example:

```py
import pyvista as pv
import numpy as np

N = 0.5
x = np.arange(0, 10, N)
y = np.arange(0, 10, N)
z = np.arange(0, 10, N)
x, y, z = np.meshgrid(x, y, z)

def density_func(xx, yy, zz):
    rho_w = 997 # kg / m^3
    return np.sin(xx) * np.sin(yy) * rho_w + zz

p = density_func(x,y,z)

mesh = pv.StructuredGrid(x, y, z)
mesh['p'] = p.ravel(order='f')

mesh.contour().plot(scalars='p', notebook=0)
```

<img width="624" alt="Screen Shot 2021-11-30 at 11 05 06 PM 1" src="https://user-images.githubusercontent.com/22067021/144180819-a089e0e9-570c-4afb-995b-dd73cd9cbade.png">

**Reply by Negar-Erfanian** [reply]
Thanks for your reply!

Here is my code and attached are the photos from this code. I actually need to add transparency but as I work with opacity (no matter what number I assign to it) my figures disappear completely. I actually need help soon as I need these figures for less than a day. Would you please help me with this? Thank you very much.


    N = 25
    from pyvista import examples
    from matplotlib.cm import get_cmap
    tex = examples.download_masonry_texture()
    
    
    
    pv.global_theme.transparent_background = False
    pv.global_theme.show_edges = False
    pv.use_transparency=True


    x = np.linspace(BBOXES[dataset_name][0], BBOXES[dataset_name][1], N)
    y = np.linspace(BBOXES[dataset_name][2], BBOXES[dataset_name][3], N)
    z = np.linspace(BBOXES[dataset_name][4]/30, BBOXES[dataset_name][5], N)

    s = np.stack([x, y, z], axis=1)
    
    
    

    X, Y, Z1 = np.meshgrid(s[:, 0], s[:, 1], s[:,2])            #(N,N,N) for each
    
    S = np.stack([X.reshape(-1), Y.reshape(-1), Z1.reshape(-1)], axis=1)             #(N**3,3)
    S = torch.tensor(S).to(device)
    S = S.double() if fp64 else S.float()

    S = (S - S_mean.to(S).T) / S_std.to(S).T
    logp = loglik_fn(S)            #(N**3,1)
    

    Z = logp.exp().detach().cpu().numpy().reshape(N, N, N)

    
    mesh = pv.StructuredData(X, Y, Z1)
    
    mesh['Z'] = Z.ravel(order='f')
    
    #tex = examples.download_rgba_texture()

    tex = examples.download_masonry_texture()

    mesh.texture_map_to_plane(inplace=True)

    p = pv.Plotter(shape=(1, 2))
    p.subplot(0, 0)
    p.add_text('pic1')
    p.add_mesh(mesh, cmap="viridis",lighting=True, scalars='Z',  color=True)
    
    p.subplot(0, 1)
    p.add_text('pic2')
    p.add_mesh(mesh.contour(), texture = tex, scalars='Z',  color=True)
    
    p.show()


<img width="1368" alt="Screen Shot 2021-12-02 at 7 30 14 PM" src="https://user-images.githubusercontent.com/33331012/144529425-8bb788a8-6024-4454-9db7-320540c4b87f.png">

**Reply by akaszynski** [reply]
I'd go with using a [UniformGrid](https://docs.pyvista.org/examples/00-load/create-uniform-grid.html) as you can use [add_volume](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_volume.html).

Here's @banesullivan's example adapted to use a [pyvista.UniformGrid](https://docs.pyvista.org/api/core/_autosummary/pyvista.UniformGrid.html)
```py
import pyvista as pv
import numpy as np

grid = pv.UniformGrid((20, 20, 20))
grid.spacing = (0.5, 0.5, 0.5)

def density_func(xx, yy, zz):
    rho_w = 997 # kg / m^3
    return np.sin(xx) * np.sin(yy) * rho_w + zz

grid['p'] = density_func(grid.x, grid.y, grid.z)/4

p = pv.Plotter()
p.add_volume(grid, cmap='blues', show_scalar_bar=False)
p.show()
```
![tmp](https://user-images.githubusercontent.com/11981631/144530844-399a5443-5cdb-430b-a30b-db03806e67cd.png)

**Reply by Negar-Erfanian** [reply]
Thanks for your reply. The problem is not with spacing. Actually, pic2 is the contour still using structureddata grid, which contains spacing by itself. I want to see between the layers in the contour (transparent) or the whole space with no contouring but still transparent.

**Reply by banesullivan** [reply]
> or the whole space with no contouring but still transparent.

This is what a `UniformGrid` and volume rendering would provide.

Otherwise, pass an opacity when plotting the contours to see the different layers much like this example: https://docs.pyvista.org/examples/01-filter/interpolate.html#complex-interpolation

![sphx_glr_interpolate_004](https://user-images.githubusercontent.com/22067021/144539664-aad02db4-9fba-4e63-bea9-4bcdab65e4f6.png)

**Reply by Negar-Erfanian** [reply]
Dear @banesullivan and @akaszynski 
Thanks for your answers.

@banesullivan :
Let's get back to the example you provided before.
Why does the plot disappear as I add opacity?

import pyvista as pv
import numpy as np

N = 0.5
x = np.arange(0, 10, N)
y = np.arange(0, 10, N)
z = np.arange(0, 10, N)
x, y, z = np.meshgrid(x, y, z)

def density_func(xx, yy, zz):
    rho_w = 997 # kg / m^3
    return np.sin(xx) * np.sin(yy) * rho_w + zz

p = density_func(x,y,z)

mesh = pv.StructuredGrid(x, y, z)
mesh['p'] = p.ravel(order='f')

mesh.plot(scalars='p', notebook=0,opacity = 0.9)

**Reply by RichardScottOZ** [reply]
You have a Structured Grid there - they are using a Uniform Grid, which gives them other plotting capabilities?

**Reply by Negar-Erfanian** [reply]
So basically opacity does not work with the structuredgrid...

**Reply by banesullivan** [reply]
Opacity works on as expected on `StructuredGrid`s -- its just that you only see the outer faces when rendering a `StructuredGrid`. In order to see inside the mesh, you need to perform volume rendering which is only supported for `UniformGrid`s at this time. So I would use @akaszynski's example above

**Reply by banesullivan** [reply]
You can see the difference in my [comment](https://github.com/pyvista/pyvista/discussions/1904#discussioncomment-1741571) where the left is volume rendering and on the right you have to take contours to produce isosurfaces which are made transparent

**Reply by Negar-Erfanian** [reply]
Thank you all for your answers. 
The problem with @akaszynski's example using uniformgrid and volume rendering is the density function having the grid.x, grid.y, and grid.z as the inputs whereas in my specific code I need to input x, y and z as is and not their grid form.
As an example, if my x ranges from 100 to 150, the way this example inputs grid.x to the density function is the values between 0 and 10 and not between 100 and 150 as needed. 

I was thinking of first outputting the density function using x, y and z (and not grid.x, grid.y and grid.z) and maybe doing something with this output to be able to plot it using volume rendering, but I don't really know how to do it.

In the given example by @banesullivan the data is type polydata which at some point I also thought of maybe converting my numpy array to polydata? but my output is (N,N,N) and in order to convert to polydata I think my output should be in the form of (N,3)?

Any insights would be very much appreciated!!!

**Comment by RichardScottOZ** [comment]
You can use those numbers in your grid...and yes for polydata.
